### DISCUSSION TITLE
ldc2 shortened methods

### DISCUSSION THREAD

**Discussion by VadimkaG** [dubinc/dub#1196] [Q&A]
Tell me what I'm doing wrong?

`$ cat dub.json`
```json
{
	"authors": [
		"Vadimka"
	],
	"copyright": "Copyright © 2024, Vadimka",
	"dependencies": {
		"godot-dlang": "~>0.7.0"
	},
	"description": "godot game",
	"dflags": [
		"-preview=shortenedMethods"
	],
	"license": "proprietary",
	"name": "battlenauts",
	"targetType": "dynamicLibrary"
}
```
`$ dub build --force`
```
Performing "debug" build using /usr/bin/ldc2 for x86_64.
godot-dlang:util 0.7.0: building configuration "library"...
utf_bc 0.2.1: building configuration "library"...
godot-dlang 0.7.0: building configuration "default"...
Running pre-build commands...
../../.dub/packages/godot-dlang-0.7.0/godot-dlang/src/godot/variant.d(716,43): Error: => shortened method not enabled, compile with compiler switch `-preview=shortenedMethods`
../../.dub/packages/godot-dlang-0.7.0/godot-dlang/src/godot/variant.d(718,43): Error: => shortened method not enabled, compile with compiler switch `-preview=shortenedMethods`
/usr/bin/ldc2 failed with exit code 1.
```
`$ ldc2 --version`
```
LDC - the LLVM D compiler (1.30.0):
  based on DMD v2.100.1 and LLVM 14.0.6
  built with LDC - the LLVM D compiler (1.30.0)
  Default target: x86_64-pc-linux-gnu
  Host CPU: goldmont
```

**Comment by selcukcukur** [comment]
@VadimkaG The error you’re hitting isn’t about your `dub.json` itself — the problem is that the `godot-dlang` package uses the **shortened methods preview feature** (`=>` shorthand), but your build isn’t actually compiling with that preview flag enabled.

A couple of key points:

- In `dub.json`, you added `"dflags": ["-preview=shortenedMethods"]`.  
- However, `dub` only applies `dflags` to your own target. Dependencies (like `godot-dlang`) don’t inherit those flags automatically.  
- That’s why you see the error inside `godot/variant.d`: it’s being compiled without the preview flag.

### How to fix it
You need to make sure the preview flag is passed to **all packages**, not just your own:

1. **Use `buildRequirements` in dub.json**  
   ```json
   "buildRequirements": ["previewShortenedMethods"]
   ```
   This tells dub to apply the preview feature globally, including dependencies.

2. **Or set `DFLAGS` environment variable** before building:  
   ```bash
   export DFLAGS="-preview=shortenedMethods"
   dub build --force
   ```
   That way every compilation unit gets the flag.

3. **Alternative:** If you only want it for your project, you’d need to patch `godot-dlang` to add the flag in its own package config, but usually the global requirement is simpler.
